### PR TITLE
Fixed "Segfault being dragged by nursebot in build 9424 #32929"

### LINKED ISSUE #32929: Segfault being dragged by nursebot in build 9424
# Describe the bug

I was going through an ice lab and came across a nursebot that actually appeared to be aggro, unlike other nursebots I've seen, and the log indicated that "the nurse bot scans you and seems to detect something". I was curious, so I went up to the bot, which then grabbed me and started dragging me in the direction of the autodoc. Not wanting to lose this character in case it vivisected me or whatever, but wanting to see what happened, I saved and exited to make a backup of the save. Upon loading the save now, it consistently crashes with a segfault as soon as I disable safe mode and do anything. Passing time, trying to move away, or attacking the nursebot all cause a crash.

I tried debug teleport, which just added a couple lesser errors before the segfault, but then thankfully debug "kill all monsters" destroyed the nursebot and prevented the segfault. So if all else fails, that will work around it.

This save was started several version back, with me occasionally updating, so it's possible that's related. I have been playing on this version for a couple hours with no real incident, however.

# Steps To Reproduce

Possibly the above with having a nursebot decide to drag you, saving, then loading the save and continuing. Or it might be something specific to my save such as mods, I haven't tried reproducing it with a fresh unmodded game.

# Expected behavior

No crash.

# Screenshots

**If applicable**, add screenshots to help explain your problem.

cdda_segfault_next_turn

# Versions and configuration

- OS: Windows 10
    - OS Version: 10.0 1903
- Game Version: 0.D-6153-g89a33be [64-bit]
- Graphics Version: Tiles
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    More Locations [more_locations],
    Urban Development [Urban_Development],
    Fuji's More Buildings [FujiStruct],
    Folding Parts pack [deoxymod],
    More Survival Tools [More_Survival_Tools],
    Alternative Map Key [alt_map_key],
    Mutant NPCs [mutant_npcs],
    No Filthy Clothing [no_filthy_clothing],
    Prevent Zombie Revivication [no_reviving_zombies]
]

# Additional context

Logs:
[cdda_segfault_logs.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/3465901/cdda_segfault_logs.zip)
The save file was a bit big, so I uploaded to Mega:
https://mega.nz/#!Z8xWFCLB!c4wQw_vxFJrGzsl8UyV7Z5Yw7AwMQaZQwOKaftSfJoc

### PR DESCRIPTION
**Pull request by ipcyborg**
#### Summary
SUMMARY: Bugfixes "Segfault being dragged by nursebot in build 9424 #32929"

#### Purpose of change
Fixes #32929

#### Describe the solution
Crash was caused by not initialized pointer `dragged_foe` in `monster`. It was not properly saved/loaded. 
Fixed by using saving/loading by ID (added `dragged_foe_id`). When `dragged_foe` is required, it is determined from `dragged_foe_id`, only once.
The crash in saved game provided in #32929 is fixed, but the dragging process is interrupted after the game loading: no valid `dragged_foe_id` was saved.  Still no crash. 
Then I have tested saving when dragged and the loading - works fine, no dragging interruption.

#### Describe alternatives you've considered
I'm not sure if this logic of saving and deferred loading of pointer by ID is correct. 
May be I can just use the ID to immediately get a pointer during the loading process? Though I really doubt this.

#### Additional context
A lot of style changes. Should I just astyle a changed part of code instead of all changed files?

### CODE REVIEW

**Comment by ZhilkinSerg** [comment]
Hmm... Something is not right with astyle.

**Comment by ipcyborg** [comment]
I'm using astyle 3.0.1.
Command line arguments are from here:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/CODE_STYLE.md

What am I doing wrong? :)

File: src/monster.h [hunk 1]
```diff
@@ -420,6 +420,7 @@ class monster : public Creature
         int wandf;           // Urge to wander - Increased by sound, decrements each move
         std::vector<item> inv; // Inventory
         player *dragged_foe; // player being dragged by the monster
+        int dragged_foe_id; // id of player being dragged by the monster (for save/load)
```
[the remark this answers is not in the document]
Comment (ipcyborg) [reply src/monster.h]: Thanks!
Fixed in that branch. 
Or should I create a separate PR for that?

[the remark this answers is not in the document]
Comment (Qrox) [reply src/monster.h]: A separate PR is needed to merge your change into master

File: src/savegame_json.cpp [hunk 2]
```diff
@@ -1886,6 +1889,10 @@ void monster::store( JsonOut &json ) const
     }
     json.member( "inv", inv );
 
+    if( dragged_foe ) {
+        json.member( "dragged_foe_id", dragged_foe->getID() );
```
Comment (Qrox) [inline src/savegame_json.cpp]: I haven't tested, but I suspect the cause of #33051 might be that the object dragged_foe points to has already been destructed here, leading to a use-after-free problem. It's probably better to always store dragged_foe_id in `monster` and retrieve the pointer on demand, which should eliminate the crash. (This is also how passenger saving/loading are done in the vehicle code, if I understood the code correctly).

File: src/monster.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (Qrox) [reply src/monster.h]: Actually it seems that, from the comments of `player::setID()` in player.h and the initial value of `game::next_npc_id` (which is 1), the default value for the id should be 0, and it's `player::setID()` that needs fixing.

[the remark this answers is not in the document]
Comment (ipcyborg) [reply src/monster.h]: Will be fixed by #33145.

**Pull request merged**